pystruct/inference/inference_methods.py

```diff
@@ -350,30 +378,175 @@ def inference_ad3(unary_potentials, pairwise_potentials, edges, relaxed=False,
+def inference_ad3plus(l_unary_potentials, l_pairwise_potentials, l_edges
+                      , relaxed=False
+                      , verbose=0, return_energy=False, branch_and_bound=False
+                      , constraints=None, inference_exception=None):
+    """
+    Inference with AD3 dual decomposition subgradient solver.
+
+    Parameters
+    ----------
+    unary_potentials : nd-array, shape (n_nodes, n_states)
+        Unary potentials of energy function.
+
+    pairwise_potentials : nd-array, shape (n_states, n_states) 
+                        or (n_states, n_states, n_edges).
+        Pairwise potentials of energy function.
+        If the first case, edge potentials are assumed to be the same for all 
+        edges.
+        In the second case, the sequence needs to correspond to the edges.
+
+    edges : nd-array, shape (n_edges, 2)
+        Graph edges for pairwise potentials, given as pair of node indices. As
+        pairwise potentials are not assumed to be symmetric, the direction of
+        the edge matters.
+
+    relaxed : bool (default=False)
+        Whether to return the relaxed solution (``True``) or round to the next
+        integer solution (``False``).
+
+    verbose : int (default=0)
+        Degree of verbosity for solver.
+
+    return_energy : bool (default=False)
+        Additionally return the energy of the returned solution (according to
+        the solver).  If relaxed=False, this is the energy of the relaxed, not
+        the rounded solution.
+
+    branch_and_bound : bool (default=False)
+        Whether to attempt to produce an integral solution using
+        branch-and-bound.
+
+    constraints : list of logical constraints or None (default:=None)
+        A logical constraint is tuple like 
+            ( <operator>, <unaries>, <states>, <negated> )
+        where:
+        - operator is one of:
+             'XOR' 'XOROUT' 'ATMOSTONE' 'OR' 'OROUT' 'ANDOUT' 'IMPLY'
+        - unaries is a list of the index of each unary involved in this 
+        constraint
+        - states is a list of unary states (class), 1 per involved unary. If the
+         states are all the same, you can pass it directly as a scalar value.
+        - negated is a list of boolean indicating if the unary must be negated.
+        Again, if all values are the same, pass a single boolean value instead
+        of a list 
+        
+        NOTE: this hard logic constraint mechanism has been developed for the 
+        EU project READ, by JL Meunier (Xerox), in November 2016.
+        The READ project has received funding from the European Union's Horizon
+        2020 research and innovation programme under grant agreement No 674943.
+    
+    Returns
+    -------
+    labels : nd-array
+        Approximate (usually) MAP variable assignment.
+        If relaxed=False, this is a tuple of unary and edge 'marginals'.
+
+    Code written on Feb 2017 to deal with multiple node types, by JL Meunier, 
```

Comment:
I think it's more maintainable to have these either at the top of the file in the copyright header or in the readme (where it already is). If everybody that added/changed a method would add their affiliation to the docstring the scikit-learn docs would be unreadable now.

Reply:
(I don't have a very strong opinion though).

Reply:
agree => removed